### ISSUE TITLE
Multiple-monitors support in fullscreen mode

### ISSUE THREAD

**Issue by bunnybot** [widelands/widelands#1079]
On a multiple monitor setup, Widelands always uses the primary monitor in fullscreen mode. It should be possible to select which monitor to play on in the options, ideally with an option to use all of them.

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2018-09-08T16:11:40Z
 * owner: nordfriese
 * the launchpad url was https://bugs.launchpad.net/bugs/1791443

**Comment by bunnybot** [comment]
(by gunchleoc)
I don't have a multi-monitor setup, but since we don't have any code to handle this, I'm setting the bug to confirmed.

**Comment by bunnybot** [comment]
(by franku)
I have a multimonitor setup and my display manager/window manager have always trouble with it :-D

The problem here is that it is a problem of window/display manager configuration in conjunction with the graphics adapter/driver, not a problem of a single program.

> ideally with an option to use all of them.

This depends also to the window manager and graphics driver, not to a program. Example: https://www.facebook.com/photo.php?fbid=1386381771377233&set=o.326816027402782&type=3&theater

What i want to say: There are so many possibilities, also depending on the operating system and the graphics driver, that it will be impossible to achieve a solution for widelands to this bug.

I would vote to mark this bug as invalid or won't fix.

**Comment by bunnybot** [comment]
(by nordfriese)
I don´t think it only depends on the graphic driver and window manager. The default behavior on my system is that a fullscreen window is placed on the monitor it was located on before maximizing. Widelands already has a different behavior, so I´d take a wild guess that the SDL can have some influence here.
Other programs have other handling methods. For example, OpenTTD allows me to choose a resolution equal to the combined size of both monitors, and when entering fullscreen mode at that res, the game uses both monitors. That´s what I would love to have for Widelands.
I don´t know the SDL very well, but it might be worth to check whether it allows to implement such a behavior.

**Comment by bunnybot** [comment]
(by gunchleoc)
SDL does have some functions for listing the monitors, so we should be able to take it from there and at least improve the situation - staying on the same monitor while switching the mode would be good.

Code snippet for iterating through the available monitors:

https://wiki.libsdl.org/SDL_GetCurrentDisplayMode

**Comment by jmoerschbach** [comment]
Hi,

is this still a problem? At least on my machine (Windows 10, two identical monitors with 1920x1080) switching between fullscreen and window mode lets widelands stick to the same monitor...

Is this different on Linux with various window managers?

However, Widelands ALWAYS starts on primary monitor, regardless on which monitor the console from where I start widelands from is located...

**Comment by Noordfrees** [comment]
My second monitor is defective so I can't test this currently. But the last behaviour I remember was:
- When Widelands starts in windowed mode, it starts on the monitor on which the mouse pointer is currently located. (This behaviour is consistent with all other programs.)
- When Widelands starts in fullscreen mode, it always uses the primary monitor (bug, most programs don't do this)
- Switching from fullscreen to windowed keeps WL on the same screen (as it should)
- Switching from windowed to fullscreen it *sometimes* uses the current and *sometimes* the primary monitor (bug)

This was observed on Ubuntu MATE 19.04 64 bit with the Marco window manager (Composite enabled).

**Comment by Noordfrees** [comment]
Oh and I *believe* (not 100% certain though) that when I first reported this I was using plain Ubuntu (probably 18.04, 32 bit) with the GNOME Flashback desktop environment and the Metacity window manager, so this was observed with multiple DEs and WMs.

**Comment by klaus-halfmann** [comment]
OS OSX is starts on the main Monitor, but using ctrl^F and moving it to some other Monitor works just fine. No Issue for me.

**Comment by jmoerschbach** [comment]
> 
> 
> My second monitor is defective so I can't test this currently. But the last behaviour I remember was:
> 
>     * When Widelands starts in windowed mode, it starts on the monitor on which the mouse pointer is currently located. (This behaviour is consistent with all other programs.)
>
>     * When Widelands starts in fullscreen mode, it always uses the primary monitor (bug, most programs don't do this)
can't confirm, widelands always starts on the primary monitor, regardless of fullscreen or window mode. Also, at least on windows, programs do start on the primary screen even when the cursor is on the secondary screen...

>     * Switching from fullscreen to windowed keeps WL on the same screen (as it should)
> 
>     * Switching from windowed to fullscreen it _sometimes_ uses the current and _sometimes_ the primary monitor (bug)

can't confirm. It works perfectly well for me. As @klaus-halfmann does not have any issues on MacOSX can someone with a multi monitor setup try on linux once again? :)
> 
> This was observed on Ubuntu MATE 19.04 64 bit with the Marco window manager (Composite enabled).

**Comment by Niektory** [comment]
On my system (Manjaro Linux / Xfce 4.14) as of 4b60d625701eaa48ed114484c6c9f05889fff197 it works as follows:
- Widelands always starts on the primary monitor, regardless of windowed / fullscreen mode.
- Switching from windowed to fullscreen or vice versa keeps Widelands on the same screen it was at before switching.
- It doesn't matter where the mouse cursor is at.
- Most other programs work differently: some of them start on the screen where the mouse cursor is at, and some others remember the screen and position they were at when they were last closed and open there.

**Comment by frankystone** [comment]
> On my system (Manjaro Linux / Xfce 4.14) as of [4b60d62](https://github.com/widelands/widelands/commit/4b60d625701eaa48ed114484c6c9f05889fff197) it works as follows:
> 
> * Widelands always starts on the primary monitor, regardless of windowed / fullscreen mode.
> * Switching from windowed to fullscreen or vice versa keeps Widelands on the same screen it was at before switching.
> * It doesn't matter where the mouse cursor is at.
> * Most other programs work differently: some of them start on the screen where the mouse cursor is at, and some others remember the screen and position they were at when they were last closed and open there.

This is exactly the behavior on my system: Archlinux with KWin (KDE/Plasma) as windowmanager.

Don't know how the switch to wayland may affect this bug in the future.

**Comment by klaus-halfmann** [comment]
Saving the screen position can only be done if we validate them before restoring.
e.g. I start a game with my dual/triple Monitor setup, but late continue with only my
Laptop Monitor, without check the window would be offscreen and the user left without clue.

**Comment by Noordfrees** [comment]
So I summarize:

- The behaviour on multiple monitor setups is highly platform-specific
- Therefore it might be safer not to attempt to change that
- I still think it would be great to have an option to extend WL over *all* monitors in fullscreen mode
- We could then also consider allowing users to configure the extra screens for special tasks, e.g. one independent viewport per monitor

**Comment by Niektory** [comment]
Possible improvements to fullscreen handling on multiple monitors – do we want any of these?
1. Remember the last window position. Now if one prefers to play on a secondary monitor they have to move the window there every time they start Widelands. This is especially cumbersome if using fullscreen.
2. (Optionally?) Disable "minimize on focus loss", so you can do something on another monitor and keep Widelands on screen when in fullscreen mode. Currently you can get this behavior by setting the `SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0` environment variable or by using the fullscreen function of your window manager.
3. An option to choose the fullscreen monitor from the options screen.

As for using all monitors at once, I investigated how OpenTTD handles it. It can use both SDL2 and Allegro 4.
- When using SDL2 it only lists resolutions supported by the primary monitor, similar to Widelands.
- When using Allegro 4, enabling fullscreen takes over all monitors and a resolution corresponding to the entire space spanning all monitors can be selected.

In SDL2 fullscreen means one monitor only. It's possible to disable some features so it can't detect separate monitors anymore but it seems like a bad idea (and worked badly when I did a quick test).

So if we wanted to play Widelands on multiple monitors using SDL2 I can think of several possible approaches, ordered from the simplest to the most ambitious:
1. Give up on fullscreen, use windowed mode and stretch the window over multiple monitors. Already possible with #4125.
2. Use windowed mode and some tricks to fake fullscreen: move and resize to cover all monitors, remove window borders, enable "always on top", minimize on focus loss.
3. Have multiple fullscreen windows – one for each monitor – and add an abstraction layer that combines the input events and splits the output surface, so the game can treat them like one big window.
4. Have any number of windows that can be all moved/resized/fullscreened independently. One of them would be "primary" and display the main game, the others would display additional viewports and whatnot.

**Comment by gunchleoc** [comment]
I have never used a multi-monitor setup and I have none available, so I don't have any informed opinions on this.

**Comment by frankystone** [comment]
> 1. Remember the last window position. Now if one prefers to play on a secondary monitor they have to move the window there every time they start Widelands. This is especially cumbersome if using fullscreen.
> 2. (Optionally?) Disable "minimize on focus loss", so you can do something on another monitor and keep Widelands on screen when in fullscreen mode. Currently you can get this behavior by setting the `SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0` environment variable or by using the fullscreen function of your window manager.

I think both would be good.

> 3. An option to choose the fullscreen monitor from the options screen.

If 1. get implemented, this would be superfluous, imho.

> So if we wanted to play Widelands on multiple monitors using SDL2 I can think of several possible approaches, ordered from the simplest to the most ambitious:
> 
> 2. Use windowed mode and some tricks to fake fullscreen: move and resize to cover all monitors, remove window borders, enable "always on top", minimize on focus loss.

This sounds very tricky to me. It may can lead into some bugs if involved third party software (SDL, window manager) get updated?

> 3. Have multiple fullscreen windows – one for each monitor – and add an abstraction layer that combines the input events and splits the output surface, so the game can treat them like one big window.

Sounds complicated...

> 4. Have any number of windows that can be all moved/resized/fullscreened independently. One of them would be "primary" and display the main game, the others would display additional viewports and whatnot.

If i understand correctly this means i can place any window (e.g. inventory, statistics, watchwindows, ...) outside the main view? So i would have the mainview on monitor1 in fullscreen, but can move any subwindow on monitor2?

**Comment by klaus-halfmann** [comment]
I have a retina Display and can attach a 4k and a Full  HD Monitor. Playing e.g. crater on 4k, I can see the complete Map and some windows without scrolling. SDL cooperates nicely with macOS on resizing the screen.  #4125 does **not** allow e to strech the window across multiple screens, but thats a macOS "feature" I assume.  I am quite happy wth the situation as is (assuming #4125 gets merged.) I assume SDL can not help us with multiple monitors. 

All the ideas 1-4 give me more or less a headache. Im am not deep inough into our homemade window handling to have any idea how much effort this is. So I would put this on hold until we have a sloid Idea how to do this.

**Comment by Niektory** [comment]
> > 2. Use windowed mode and some tricks to fake fullscreen: move and resize to cover all monitors, remove window borders, enable "always on top", minimize on focus loss.
>
> This sounds very tricky to me. It may can lead into some bugs if involved third party software (SDL, window manager) get updated?

It could be fragile, yes.

> > 4. Have any number of windows that can be all moved/resized/fullscreened independently. One of them would be "primary" and display the main game, the others would display additional viewports and whatnot.
>
> If i understand correctly this means i can place any window (e.g. inventory, statistics, watchwindows, ...) outside the main view? So i would have the mainview on monitor1 in fullscreen, but can move any subwindow on monitor2?

Possibly. I don't have a solid idea how the secondary windows would work or what they would contain, just mentioning it as a possibility. I doubt we'll go for it anyway.

**Comment by bkil** [comment]
Let me share another data point on a multi-monitor setup. If you have one screen with a smaller resolution and another one with a bigger resolution, I can only select a resolution up to what the smaller (primary one) supports, while I would expect to be able to choose a higher resolution (or just type it in myself).

I prefer to play on the larger screen, while running the terminal, manual, chat, music and other apps on the smaller one, hence I don't want to switch it off just to work around this.

**Comment by frankystone** [comment]
@bkil : I do have also the same setup two screens with different resolutions. Handling of the window is much improved in version 1.0. Now i can start widelands in window mode, if on wrong screen move it to the correct screen, and hit Strg-f ( or hit the "Maximize"-button) to switch to full screen.
In window mode it is also possible to stretch the window over both screens.

**Comment by frankystone** [comment]
Just stumbled over this issue... is this still an issue since one can stretch the widelands window over more than one monitor?

I am in favor to close this issue.

**Comment by bkil** [comment]
Alright, it works now, thanks.

**Comment by frankystone** [comment]
Closing ....
